### PR TITLE
Host and Query an Encrypted model

### LINKED ISSUE #207: hosting an encrypted model on a grid worker
Part of Epic: #203 
--

**Blocked by: #204**

Same as #204 but encrypted :).

### LINKED ISSUE #210: Querying an Encrypted Model on a Grid Worker
Part of Epic: #203 
--

**Blocked by: #207**

As a part of the EMLaaS Demo Epic (#203), we want to be able to query an encrypted model on a grid worker such that the prediction occurs entirely using Secure Multi-Party Computation.

The API should look something like:

```python
pointer_to_encrypted_model = grid_client.models['my_model_name']
dummy_input_data = th.zeros(10,10) 
prediction = pointer_to_encrypted_model(dummy_input_data)
```
Note that there's no notion of the API that is explicitly different because the model happens to be encrypted. The input data (for the prediction) should be shared automatically, the prediction executed, and the result decrypted.

**Acceptance Criteria:**
1) the API should work even if the input data isn't encrypted (but should also work if it's already MPC shared)
2) the predict functionality should have an optional parameter for not decrypting the result.

### LINKED ISSUE #211: Notebook demo for deploying and querying an encrypted hosted model
Part of Epic: #203
--

**Blocked by: #210 , #208**

A demo showing how to deploy and query an encrypted, hosted model using the skin cancer dataset.

*The Dataset:* https://github.com/dasoto/skincancer
*The Model:* https://towardsdatascience.com/skin-cancer-classification-with-machine-learning-c9d3445b2163

### PR DESCRIPTION
**Pull request by mari-linhares**
Fix #207, fix #210, fix #211 

**Merge blocked by:**

- [x] https://github.com/OpenMined/PySyft/pull/2590
- [x] https://github.com/OpenMined/PySyft/pull/2598
- [ ] Unintentionally I merged some commits from #293 here, after #293 is merged we should create a new clean PR to add host and query encrypted models.

### CODE REVIEW

**Comment by Quisher** [comment]
can we have the file extensions set to ipynb? That way the notebook can be viewed on GitHub too.

**Comment by mari-linhares** [comment]
Will do, thanks for noticing!

**Comment by mari-linhares** [comment]
@Quisher lmk if you have any thoughts on this!

File: app/websocket/app/main/events.py [hunk 1]
```diff
@@ -43,8 +43,13 @@ def cmd(message):
...
         worker = hook.local_worker
-        if not len(worker.current_objects()):
-            recover_objects(hook)
+        # mari-linhares: I've commented these lines
+        # due to issues storing plans on the database.
+        # Maybe there are too many bytes being stored
+        # but more exploration is needed.
+        # TODO: uncomment these lines
+        # if not len(worker.current_objects()):
+        #     recover_objects(hook)
```
Comment (mari-linhares) [inline app/websocket/app/main/events.py]: @IonesioJunior any suggestions on how to deal with this? I think we can just check the types before storing in the database, what do you think?

Comment (IonesioJunior) [reply app/websocket/app/main/events.py]: sounds a good idea to me

File: grid/client.py [hunk 2]
```diff
@@ -84,10 +84,51 @@ def models(self, N: int = 1):
...
+
+        A wrapper for the send operation.
+
+        This method is used when a user wants to save the model on the worker,
+        but does not intend to serve it using a Rest API due to privacy reasons.
+        """
+        return model.send(self)
+
+    def serve_model(self, model, model_id=None):
         # If the model is a Plan we send the model
```
Comment (Quisher) [inline grid/client.py]: have separate functions for both? or multiple parameters? Doesn't this increase confusion.

Comment (mari-linhares) [reply grid/client.py]: Yes, I was thinking about this as well, will merge them into a single function is less confusing.

File: grid/client.py [hunk 3]
```diff
@@ -104,38 +104,81 @@ def destroy(self):
...
+        Args:
+            model: A jit model or Syft Plan.
+            model_id: An integer or string representing the model id used to retrieve the model
+                later on using the Rest API. If this is not provided and the model is a Plan
+                we use model.id, if the model is a jit model we raise an exception.
+            is_private_model: A boolean indicating if the model is private or not. If the model
+                is private the user does not intend to serve it using a Rest API due to privacy reasons.
+
+        Returns:
+            None if is_private_model is True, otherwise it returns a json object representing a Rest API response.
```
Comment (IonesioJunior) [inline grid/client.py]: Couldn't we return an unified type of result? Maybe we can handle the different responses (websocket /rest) inner the method and return a unified result.

Comment (IonesioJunior) [reply grid/client.py]: Just an idea :).

Comment (mari-linhares) [reply grid/client.py]: I agree, we should definitely standardize return types. But I think we could do this in a different PR, this is already bigh enough. Is this okay?

Comment (IonesioJunior) [reply grid/client.py]: Ok.

**Pull request merged**